File: etc/.buildTrie.go
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"os"
	"sort"
	"strings"
	"sync"
)

type trie struct {
	sync.Mutex
	t map[string]*trie
	w []string
}

var (
	queue = make(chan string)
	workers = flag.Int("workers", 4, "")
	indent = flag.Bool("indent", false, "")
	stop = flag.Int("stop", 0, "")
	wg = &sync.WaitGroup{}
)

func main() {
	flag.Parse()
	t := trie{}
	t.consumeQueue()
	log.Println("building")
	num := t.build()
	log.Println("trimming")
	t.trim()
	log.Println("done")
	c := t.count("")
	if c == num {
		log.Println("check passed: ", num)
	} else {
		log.Fatalf("failed check: %d/%d", c, num)
	}
	t.write()
	log.Println("closing")
}

func (t *trie) consumeQueue() {
	for i := 0; i < *workers; i++ {
		wg.Add(1)
		go t.work()
	}
}

func (t *trie) build() int {
	seen := make(map[string][]string)
	num := 0
	skipped := 0
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		word := scanner.Text()
		if handle(word, seen) {
			skipped++
		}

		num++
		if num % 1000 == 0 {
			log.Printf("%s\t%d/%d\t%0.2f%%\n", word, skipped, num, 100*float64(skipped)/float64(num))
		}
		if num == *stop {
			break
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	close(queue)
	wg.Wait()

	for sortedLetters, anagrams := range seen {
		if len(anagrams) > 1 {
			t.setAnagrams(sortedLetters, anagrams)
		}
	}

	return num
}

func (t *trie) trim() {

}

func (t *trie) count(key string) int {
	if t == nil {
		return 0
	}
	if t.t == nil {
		return len(t.w)
	} else {
		c := 0
		for rKey, r := range t.t {
			c += r.count(key+rKey)
		}
		return c
	}
}

func (t *trie) write() {
	var b []byte
	var err error
	if (*indent) {
		b, err = json.MarshalIndent(t, "", "  ")	
	} else {
		b, err = json.Marshal(t)
	}
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(string(b))
}

func (t *trie) work() {
	for word := range queue {
		letters := make(map[string]int)
		for _, letterR := range word {
			letters[string(letterR)]++
		}
		t.add(letters, word, "")
	}
	wg.Done()
}

func handle(word string, seen map[string][]string) bool {
	s := strings.Split(word, "")
	sort.Strings(s)
	sortedLetters := strings.Join(s, "")
	if anagrams, ok := seen[sortedLetters]; ok {
		seen[sortedLetters] = append(anagrams, word)
		// log.Printf("skip\t%s\t%s\n", word, anagrams[0])
		return true
	} else {
		seen[sortedLetters] = []string{word}
		queue <- word
		return false
	}
}

func (t *trie) add(letters map[string]int, word string, previous string) {
	if len(letters) == 0 {
		t.Lock()
		t.w = []string{word}
		t.Unlock()
		return
	}
	t.Lock()
	if t.t == nil {
		t.t = make(map[string]*trie)
	}
	t.Unlock()
	for letter := range letters {
		if letter < previous {
			t.Lock()
			t.t[letter] = nil
			t.Unlock()
		} else {
			rLetters := make(map[string]int)
			for rLetter, count := range letters {
				if rLetter == letter {
					if count > 1 {
						rLetters[rLetter] = count-1
					}
				} else {
					rLetters[rLetter] = count
				}
			}
			t.Lock()
			subT := t.t[letter]
			if subT == nil {
				subT = &trie{}
				t.t[letter] = subT
			}
			t.Unlock()
			subT.add(rLetters, word, letter)
		}
	}
}

func (t *trie) setAnagrams(sortedLetters string, anagrams []string) {
	if len(sortedLetters) == 0 {
		t.w = anagrams
	} else {
		key := string(sortedLetters[0])
		remainder := sortedLetters[1:len(sortedLetters)]
		t.t[key].setAnagrams(remainder, anagrams)
	}
}

func (t *trie) MarshalJSON() ([]byte, error) {
	var obj interface{}
	if t != nil {
		if t.w != nil {
			obj = t.w
		} else {
			obj = t.t
		}
	}
	return json.Marshal(obj)
}

File: etc/buildTrie.py
import collections
import json
import os
import sys
import time

sizeCutoff = 1000000000
MAGIC = "-"
targetCount = collections.defaultdict(int)
sourceToTarget = {}

def main():
	sowpods = open(sys.argv[1])
	words = json.load(sowpods)
	fh = open(sys.argv[2], 'w') if len(sys.argv) > 2 else None
	p('building')
	global bigD
	bigD = trie()
	num = build(bigD, words)
	p('built')
	numNodes = -1 # getNumNodes(bigD)
	p('trimming alts %d' % numNodes)
	trimAlts(bigD)
	numNodes = getNumNodes(bigD)
	p('done %d' % numNodes)
	c = count(bigD)
	if c == num:
		p('check passed: %d' % num)
	else:
		p('check failed: %d/%d' % (c, num))
		exit(1)
	bigD = ddToD(bigD)
	p('ddToD')
	dump(bigD, fh)
	p('closing')

def p(string):
	print time.ctime(), string

def trie():
	return collections.defaultdict(trie)

def build(d, words):
	seen = {}
	num = 0
	skipped = 0
	words = sorted(words, lambda a,b: len(a)-len(b))
	for word in words:
		if word:
			word = word.upper()
			num += 1
			if num % 1000 == 0: p('\t%s %d\t%d/%d\t%0.2f%%' % (word, len(word), skipped, num, 100*float(skipped)/num))
			sortedLetters = ''.join(sorted(word))
			if sortedLetters in seen:
				skipped += 1
				# p('skip\t%s\t%s' % (word, seen[sortedLetters]))
				addAnagram(sortedLetters, word)
			else:
				seen[sortedLetters] = word
				letters = collections.Counter(word)
				add(letters, d, sortedLetters, word)

	return num

def ddToD(d):
	d = dict(d)
	for key in d.keys():
		sub = d[key]
		if type(sub) is collections.defaultdict:
			d[key] = ddToD(sub)
	return d

def getNumNodes(d):
	numNodes = len(d)
	for letter in d:
		sub = d[letter]
		if isinstance(sub, collections.Mapping):
			numNodes += getNumNodes(sub)
	return numNodes

def trimAlts(d, chain=''):
	shouldTrim = MAGIC not in d and chain not in targetCount
	rval = shouldTrim
	for key in reversed(sorted(d.keys())):
		if key is MAGIC:
			rval = False
			continue
		sub = d[key]
		if sub is None:
			if shouldTrim:
				del d[key]
				target = sourceToTarget.pop(chain+key)
				targetCount[target] -= 1
				if targetCount[target] is 0:
					del targetCount[target]
		elif trimAlts(sub, ''.join(sorted(chain+key))):
			del d[key]
		else:
			rval = False
	return rval

def count(d, chain=''):
	if d is None:
		sub = bigD
		for letter in sorted(chain):
			if letter not in sub:
				dump(bigD, None)
				print chain, letter
				print 'failed count unreachable'
				exit(1)
			sub = sub[letter]
		return 0
	c = 0
	if not d:
		dump(bigD, None)
		print chain
		print 'failed count empty dict'
		exit(1)
	for letter in d:
		if letter is MAGIC:
			c += len(d[letter])
		else:
			c += count(d[letter], chain+letter)
	return c

def dump(d, fh):
	if fh is not None:
		json.dump(d, fh)
	else:
		print json.dumps(d, indent=2, sort_keys=True)

def add(letters, d, sortedLetters, word, previous = '', chain = ''):
	if len(letters) == 0:
		if MAGIC not in d:
			d[MAGIC] = []
		d[MAGIC].append(word)
		return

	keys = sorted(letters.keys())
	first = keys[0]
	for letter in keys:
		if letter < previous:
			if letter is first:
				d[letter] = None
				save(chain, letter)
		else:
			rLetters = dict(letters)
			rLetters[letter] -= 1
			if rLetters[letter] is 0:
				del rLetters[letter]
			sub = d[letter]
			add(rLetters, sub, sortedLetters, word, letter, chain+letter)

def save(chain, letter):
	source = chain+letter
	if source in sourceToTarget: return
	target = ''.join(sorted(chain+letter))
	sourceToTarget[source] = target
	targetCount[target] += 1


def addAnagram(sortedLetters, word):
	d = bigD
	for letter in sortedLetters:
		d = d[letter]
	d[MAGIC].append(word)

if __name__ == "__main__":
	main()
